prime-router/docs/proposals/0006-email-notification/0006-email-notification.md

```diff
@@ -0,0 +1,174 @@
+
+## Goals
+
+The EmailService should:
+
+1. Respond to known and periodic events to send templated emails
+1. Maintain a base of templates that can be updated as needs change
+1. Maintain a schedule of emails to be sent
+1. Have parameters that can be filled with the template based on
+    - most recently sent emails
+    - organization and organization participants
+    - rellevant dates
+
+## Proposal
+
+The EmailService will run as a HTTP triggered service presenting a REST-ful API to manage EmailSchedules.  It will have capabilities to LIST, CREATE, DELETE, and UPDATE EmailSchedules.  When an EmailSchedule is created, the EmailService will programatically create and launch a time-trigger based Azure function based on the 'schedule' element within the supplied message for building the desired email and notifying SendGrid.  For 'send-immediate' messages, no function will be created - but the email will be sent.
+
+The email service will store its relevant data within the Postgres database, with the exception of templates which will be maintained on the SendGrid site (with references to them in Postgres).
+
+The EmailService will take in a Schedule element that consists of a body similar to:
+``` json
+{
+    "template": "template.name", // the name of the template
+    "schedule": "chron string for scheduling", // optional, no schedule means send immediate
+    "organization" : "optional organization to send to", // ex. DHpima_az_doh, DHmt_doh, etc.
+    "recipient": [ "optional", "optional" ], //optional array of recipients emails
+    "excluding": [ "optional"], // a list of recipients to exclude, otherwise everybody within the organization receives
+    "parameters": {
+        "param1": "value for param1",
+        ...
+    }
+}
+```
+
+See [openapi.emailservice.yml](./openapi.emailservice.yml) for a detailed specification
+
+The EmailService will fire off emails based on:
+- If there are organizations, only for those organizations, otherwise, for every organization
+- If there are recipients, only for those recipients, otherwise, for every recipient, with the exception of anyone listed in "excluding"
+
+The EmailService will record all of its actions to a logging table.
+
+The EmailService will require an HttpHeader Authentication with a valid JWT token.
+
+### Special parameter values
+
+| Parameter | Description |
+| --------  | ------------|
+|$reports_since_last| retrieves an array of reports sent since the last email |
+|$today | Java Date representing the current date/time |
+|$org | The organization being referenced |
+|$user | The username being referenced |
+
+*Note: more to be added*
+
```

Comment:
It feels like everything in the organization setting should be made available. For example, state code. If a value is missing, it should be added to the organization setting. Fore example, state name.